### PR TITLE
feat: Feat/10521 react-native-performance mvp for measuring app start + JS bundle load times

### LINKED ISSUE #10521: Implement MVP for measuring app start and JS bundle load times
### What is this about?

Using the findings from the Investigation done https://github.com/MetaMask/metamask-mobile/issues/10054

Implement a new way to measure our app start times, using https://github.com/oblador/react-native-performance/blob/master/packages/react-native-performance/README.md

The two values that we would like to utilize is app start time as well as JS bundle load time. This should only exist under a DEV flag for now for debugging purposes

The purpose of this is to expose the ability for the team to begin utilizing these numbers for debugging purposes. We will follow up with using these measurements in the CI

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

- Add `react-native-performance` package as a dev dependency
- Expose both app start and JS bundle load measurement times
- Add performance measurement behind a DEV flag

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

- When debugging locally, contributors should see both an app start and JS bundle load times as a log in the terminal
- These numbers should not appear in release builds

### Stakeholder review needed before the work gets merged

- [ ] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

react-native-performance - https://github.com/oblador/react-native-performance/blob/master/packages/react-native-performance

_No response_

### PR DESCRIPTION
**Pull request by Cal-L**
## **Description**

This PR is based off of @MarioAslau 's MVP implementation of react-native-performance - https://github.com/MetaMask/metamask-mobile/pull/10560. This PR strictly focuses on only native launch and JS load times and displays the times in seconds. The performance numbers only appear in dev mode. Note: The numbers are relative to the environment that it runs in. When working on performance improvements in dev mode, we should take the baseline numbers relative to our own machines. This tool is a prerequisite for re-introducing app start test in e2e.

## **Related issues**

Fixes: #10521 

## **Manual testing steps**

- Simply run the app in dev mode and observe the performance numbers in the terminal

### **After**

image

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/core/Performance/Performance.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        // Measure app start
+        performance.measure(
+          'nativeLaunch',
+          'nativeLaunchStart',
+          'nativeLaunchEnd',
+        );
+        // Measure JS bundle load
+        performance.measure(
+          'runJsBundle',
+          'runJsBundleStart',
```
Comment (tommasini) [inline app/core/Performance/Performance.ts:29]: Curious question: why only get the entriesByName of runJsBundle and not the runJsBundleStart?

File: app/core/Performance/Performance.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    new PerformanceObserver((list) => {
+      // Get measurement entries
+      const entries = list.getEntries();
+
+      if (entries.find((entry) => entry.name === 'runJsBundleEnd')) {
+        // Measure app start
+        performance.measure(
+          'nativeLaunch',
+          'nativeLaunchStart',
```
Comment (tommasini) [inline app/core/Performance/Performance.ts:23]: Curious question: why only get the entriesByName of nativeLaunch and not the nativeLaunchStart?

File: app/core/Performance/Performance.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        performance.measure(
+          'runJsBundle',
+          'runJsBundleStart',
+          'runJsBundleEnd',
+        );
+        // Retrieve the measurements
+        const nativeLaunchEntry = performance.getEntriesByName('nativeLaunch');
+        const runJsBundleEntry = performance.getEntriesByName('runJsBundle');
+        // Get the duration
+        const nativeLaunchDuration = nativeLaunchEntry[0].duration;
```
Comment (tommasini) [inline app/core/Performance/Performance.ts:36]: Curious question: do you know why have multiple entries? And why we close the one on position 0?

File: app/core/Performance/Performance.ts (hunk 2 above)
Comment (Cal-L) [reply app/core/Performance/Performance.ts:23]: This call assigns the duration between `nativeLaunchEnd` and `nativeLaunchStart` and assigns it to the name `nativeLaunch`, which `getEntriesByName` can then access later.

File: app/core/Performance/Performance.ts (hunk 1 above)
Comment (Cal-L) [reply app/core/Performance/Performance.ts:29]: Similar answer as above

File: app/core/Performance/Performance.ts (hunk 3 above)
Comment (Cal-L) [reply app/core/Performance/Performance.ts:37]: It seems this is just the way that the data is stored, filtered, and returned

Comment (MarioAslau) [reply app/core/Performance/Performance.ts:37]: @tommasini  Cal is right, it's how they store the data.

**Review by tommasini — DISMISSED** [review]
looks awesome!
Just a small lint issue to fix!

**Review by tommasini — APPROVED** [review]

**Pull request merged**